### PR TITLE
Fix a race condition in BatchFixAllProvider's merge logic, in presence o...

### LINKED ISSUE #320: Possible bug in Batch Fixer merge algorithm
We have a code fix that uses the batch fixer to provide support for correcting all instances of a particular style problem in a document/project/solution. Even though each correction is made to a distinct `ExpressionSyntax` node in the source tree, the batch fixer fails to properly merge two changes in a particular scenario.

The problem is described (with example input/output) in the following issue:
https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/466

### PR DESCRIPTION
**Pull request by mavasani**
...f merge conflicts.

The current merge logic that merges code fixes from batch fixes ignores individual text changes that conflict with already merged batch fixes. This can be problematic if we have already merged a fix in the document in a later text span. We may end up adding text changes of kind Insert, but silently ignore text changes of kind Remove, returned by SyntaxTree.GetChanges(tree). This can cause the merged text document to be completely corrupt.
Fix is to ignore **all** the text changes from a code fix (that is part of batch fixes) if any of its text change has a conflict.

NOTE: https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/723 represents another case similar to #463, where SyntaxTree.GetChanges returns an add + remove text change with long spans, when the actual text change is just a single char replace. This causes completely independent code fixes to seem as conflicting to the batch fixer. I'll file a separate issue for this API bug.

Fixes #320

### CODE REVIEW

**Comment by mavasani** [comment]
@srivatsn @JohnHamby @tmeschter @heejaechang @shyamnamboodiripad @jmarolf can you please review?

**Comment by mavasani** [comment]
@sharwell @AArnott this should fix the document corruption from FixAll that you guys have been seeing. We might still end up missing on some fixes being merged into the final batch fix due to bugs such as #463 in SyntaxTree.GetChanges, but at least we should not see any corrupt result. Thanks again for the report!

**Comment by AArnott** [comment]
Thanks. Yes I agree skipping a fix is better than corrupting the result.

**Comment by tmeschter** [comment]
@mavasani Any good way to add a unit test for this?

**Comment by srivatsn** [comment]
:+1:

**Comment by mavasani** [comment]
@tmeschter This issue doesn't repro consistently. The batch fixer computes all the individual fixes in parallel and picks an arbitrary order to merge them. The issue only repros if a fix in a later part of a line is merged before a fix in the earlier part, where the fix in the earlier part has a conflicting text change. For the reverse merge, we would just fully ignore the subsequent fix and hence won't corrupt the resulting document.

**Comment by JohnHamby** [comment]
I sign off, because these changes produce a better situation than before, but I'm confused about at least one point. If there is a conflict, are the results deterministic, or can there be different results depending on the completion order of the merge tasks?

It's probably possible to write a useful unit test that applies a set of fixes say, 10000 times, and passes only if all attempts succeed.

**Comment by mavasani** [comment]
_"If there is a conflict, are the results deterministic, or can there be different results depending on the completion order of the merge tasks?"_

The result will be non-deterministic. We can probably tune the fixer to order fixes by span start of original diagnostic for merging them to make it deterministic. However, the original assumption was users should not be using a batch fixer if they have fixers which can produce conflicts in a fix all scenario. They should instead write up their own custom FixAllProvider. These corruption issues are showing up right now due to subtle issues in SyntaxTree.GetChanges, but once we fix this API or move to a different differ, as suggested in #470, batch fixer in merge conflict scenarios would be a low priority scenario.

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs [hunk 1]
```diff
@@ -478,6 +461,15 @@ private static Document GetReportedDocument(Diagnostic diagnostic, ImmutableDict
...
         }
 
+        /// <summary>
+        /// Try to merge the changes between <paramref name="newDocument"/> and <paramref name="oldDocument"/> into <paramref name="cumulativeChanges"/>.
+        /// If there is any conflicting change in <paramref name="newDocument"/> with existing <paramref name="cumulativeChanges"/>, then the original <paramref name="cumulativeChanges"/> are returned.
+        /// Otherwise, the newly merged changes are returned.
+        /// </summary>
+        /// <param name="oldDocument">Base document on which FixAll was invoked.</param>
+        /// <param name="newDocument">New document with a code fix that is being merged.</param>
+        /// <param name="cumulativeChanges">Existing merged changes from other batch fixes into which newDocument changes are being merged.</param>
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs]: you can use intervaltree for cumulative changes.

http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Collections/IntervalTree%25601.cs,191f1141194b2e6a

**Comment by heejaechang** [comment]
:+1:

**Pull request merged**